### PR TITLE
Support for Save & Load Modem DTLS session

### PR DESCRIPTION
**Pull request by jheiskan81**
### LwM2M client utils

- Added support for pause DTLS CID by TLS_DTLS_CONN_SAVE socket option at Rx-OFF-IDLE state and load CID socket back by TLS_DTLS_CONN_LOAD operation at Registration or De-registration state.
- New Feature can be enabled by kconfig CONFIG_LWM2M_CLIENT_UTILS_DTLS_CON_MANAGEMENT.

### LwM2M client sample

- integrated new feature with Connection Identifier overlay
- Fix overlay-aggressive-psm.conf disable eDRX when request shorter than 10.24s Active timer

### CODE REVIEW

**Review by peknis — CHANGES_REQUESTED** [review]

File: include/net/lwm2m_client_utils.h [hunk 1]
```diff
@@ -354,6 +354,20 @@ void lwm2m_utils_disable_conneval(void);
  */
 int lwm2m_utils_conneval(struct lwm2m_ctx *client, enum lwm2m_rd_client_event *client_event);
 
+/**
+ * @brief Save and load DTLS CID connection.
+ *
+ * This function should be called from an event handler registered to lwm2m_rd_client_start()
+ * before normal event handler part.
+ * he function saves the DTLS CID connection at LWM2M_RD_CLIENT_EVENT_QUEUE_MODE_RX_OFF state and
```
Comment (peknis) [inline include/net/lwm2m_client_utils.h:362]:
```suggestion
 * It saves the DTLS CID connection at LWM2M_RD_CLIENT_EVENT_QUEUE_MODE_RX_OFF state and
```

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 2]
```diff
@@ -272,6 +272,17 @@ config LWM2M_CLIENT_UTILS_DTLS_CID
 	  Requires either LWM2M_RD_CLIENT_STOP_POLLING_AT_IDLE or LWM2M_RD_CLIENT_LISTEN_AT_IDLE to be useful.
 	  This requires modem firmware 1.3.5 or newer.
 
+if LWM2M_CLIENT_UTILS_DTLS_CID
+
+config LWM2M_CLIENT_UTILS_DTLS_CON_MANAGEMENT
+	bool "Enable DTLS Connection save and load"
```
Comment (peknis) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:278]:
```suggestion
	bool "Enable DTLS connection save and load"
```

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 3]
```diff
@@ -272,6 +272,17 @@ config LWM2M_CLIENT_UTILS_DTLS_CID
...
+	depends on LWM2M_RD_CLIENT_STOP_POLLING_AT_IDLE
+	help
+	  Enable DTLS connection Identifier save at idle and load exit from idle
```
Comment (peknis) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:281]:
```suggestion
	  Enable DTLS connection identifier save at idle and load exit from idle.
```

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 4]
```diff
@@ -272,6 +272,17 @@ config LWM2M_CLIENT_UTILS_DTLS_CID
...
+	  This requires modem firmware 1.3.5 or newer.
```
Comment (peknis) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:282]:
```suggestion
	  This requires modem firmware v1.3.5 or newer.
```

**Review by peknis — CHANGES_REQUESTED** [review]

**Comment by jheiskan81** [comment]
@peknis Thanks for review.  I have fixed your change requests.

**Review by peknis — APPROVED** [review]

**Review by juhaylinen — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Review by simensrostad — APPROVED** [review]

File: samples/cellular/lwm2m_client/src/main.c [hunk 5]
```diff
@@ -385,15 +385,23 @@ static void rd_client_event(struct lwm2m_ctx *client, enum lwm2m_rd_client_event
...
-	    client_event != LWM2M_RD_CLIENT_EVENT_ENGINE_SUSPENDED) {
-		LOG_DBG("Drop network event %d at LTE offline state", client_event);
-		k_mutex_unlock(&lte_mutex);
-		return;
+	if (IS_ENABLED(CONFIG_LWM2M_CLIENT_UTILS_RAI) &&
+	    client_event == LWM2M_RD_CLIENT_EVENT_QUEUE_MODE_RX_OFF) {
+		lwm2m_rai_last();
+	}
+
+	if (IS_ENABLED(CONFIG_LWM2M_CLIENT_UTILS_DTLS_CON_MANAGEMENT)) {
```
Comment (SeppoTakalo) [inline samples/cellular/lwm2m_client/src/main.c:403]: Move this `IS_ENABLED()` into the function itself.

The main application is becoming too messy, so rather have all checks in the utility library.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_sec_con_management.c [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	} else {
+		LOG_DBG("DTLS session load ok");
+		dtls_stored = false;
+	}
+}
+
+void lwm2m_security_connection_manage(struct lwm2m_ctx *client,
+				      enum lwm2m_rd_client_event client_event)
+{
+	if ((client_event == LWM2M_RD_CLIENT_EVENT_REG_UPDATE ||
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_sec_con_management.c:63]: Check here, if we need to do anything.

Also, why is this called `lwm2m_security_`

**Comment by jheiskan81** [comment]
@SeppoTakalo Please review my changes

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 7]
```diff
@@ -270,7 +272,23 @@ config LWM2M_CLIENT_UTILS_DTLS_CID
...
+
+if LWM2M_CLIENT_UTILS_DTLS_CID
+
+config LWM2M_CLIENT_UTILS_DTLS_CON_MANAGEMENT
+	bool "Enable DTLS connection save and load"
+	depends on LWM2M_RD_CLIENT_STOP_POLLING_AT_IDLE
+	select LWM2M_CLIENT_UTILS_CONNECTION_MANAGEMENT
+	help
+	  Enable DTLS connection identifier save at idle and load exit from idle.
+	  This requires modem firmware v1.3.5 or newer.
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:285]:
```suggestion
config LWM2M_CLIENT_UTILS_DTLS_CON_SAVE
	bool "Enable DTLS connection save and load"
	depends on LWM2M_RD_CLIENT_STOP_POLLING_AT_IDLE
	help
	  Enable DTLS connection identifier save at idle and load exit from idle.
	  This requires modem firmware v1.3.5 or newer.
```

File: subsys/net/lib/lwm2m_client_utils/CMakeLists.txt [hunk 8]
```diff
@@ -23,6 +23,7 @@ zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSISTANCE
 zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_RAI
 			     lwm2m/lwm2m_rai.c)
 zephyr_library_sources(lwm2m/lwm2m_lte_notification.c)
+zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_CONNECTION_MANAGEMENT lwm2m/lwm2m_utils_con_management.c)
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/CMakeLists.txt:26]:
```suggestion
zephyr_library_sources(lwm2m/lwm2m_utils_con_management.c)
```

File: samples/cellular/lwm2m_client/src/main.c [hunk 9]
```diff
@@ -385,17 +385,16 @@ static void rd_client_event(struct lwm2m_ctx *client, enum lwm2m_rd_client_event
...
-	    client_event != LWM2M_RD_CLIENT_EVENT_ENGINE_SUSPENDED) {
+	if (client_state == LTE_OFFLINE && client_event != LWM2M_RD_CLIENT_EVENT_ENGINE_SUSPENDED) {
 		LOG_DBG("Drop network event %d at LTE offline state", client_event);
 		k_mutex_unlock(&lte_mutex);
 		return;
 	}
 
+	if (IS_ENABLED(LWM2M_CLIENT_UTILS_CONNECTION_MANAGEMENT)) {
+		lwm2m_utils_connection_manage(client, &client_event);
+	}
```
Comment (SeppoTakalo) [inline samples/cellular/lwm2m_client/src/main.c:396]:
```suggestion
	lwm2m_utils_connection_manage(client, &client_event);
```

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 10]
```diff
@@ -251,6 +251,7 @@ endif #LWM2M_CLIENT_UTILS_CELL_CONN_OBJ_SUPPORT
 
 config LWM2M_CLIENT_UTILS_RAI
 	bool "Enable release assistance indication (RAI)"
+	select LWM2M_CLIENT_UTILS_CONNECTION_MANAGEMENT
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:254]:
```suggestion
```

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 11]
```diff
@@ -259,6 +260,7 @@ config LWM2M_CLIENT_UTILS_RAI
 config LWM2M_CLIENT_UTILS_LTE_CONNEVAL
 	bool "Enable connection pre-evaluation [EXPERIMENTAL]"
 	select EXPERIMENTAL
+	select LWM2M_CLIENT_UTILS_CONNECTION_MANAGEMENT
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:263]:
```suggestion
```

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 12]
```diff
@@ -270,7 +272,23 @@ config LWM2M_CLIENT_UTILS_DTLS_CID
...
+	help
+	  Enable DTLS connection identifier save at idle and load exit from idle.
+	  This requires modem firmware v1.3.5 or newer.
+
+endif #LWM2M_CLIENT_UTILS_DTLS_CID
+
+config LWM2M_CLIENT_UTILS_CONNECTION_MANAGEMENT
+	bool
+
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:291]:
```suggestion
```

**Review by SeppoTakalo — COMMENTED** [review]
Looks otherwise fine, but this does not need extra Kconfigs.

The `lwm2m_utils_connection_manage()` pretty much selects conpile time, if there is anything to do.
Application don't need any iffdeff.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_utils_con_management.c [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+#include <net/lwm2m_client_utils.h>
+#include <lwm2m_engine.h>
+#include <lwm2m_util.h>
+#include <zephyr/net/socket.h>
+#include <zephyr/net/socket_ncs.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(lwm2m_sec_con_man, CONFIG_LWM2M_CLIENT_UTILS_LOG_LEVEL);
+
+#ifdef CONFIG_LWM2M_CLIENT_UTILS_DTLS_CON_MANAGEMENT
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_utils_con_management.c:18]:
```suggestion
```

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_utils_con_management.c [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+			       sizeof(load_dtls));
+	if (ret) {
+		ret = -errno;
+		LOG_ERR("Failed to load %d DTLS session: %d", ctx->sock_fd, ret);
+	} else {
+		LOG_DBG("DTLS session load ok");
+		dtls_stored = false;
+	}
+}
+#endif
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_utils_con_management.c:60]:
```suggestion
```

**Review by SeppoTakalo — APPROVED** [review]

**Review by peknis — APPROVED** [review]

**Pull request merged**